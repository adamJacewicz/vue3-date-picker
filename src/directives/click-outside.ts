import { Directive } from 'vue';

let onOutsideClick: (e: MouseEvent) => void;

export const clickOutside: Directive<HTMLElement> = {
  beforeMount(el, binding) {
    onOutsideClick = (e) => {
      el !== e.target && !el.contains(e.target as Node) && binding.value();
    };
    document.addEventListener('click', onOutsideClick);
  },
  beforeUnmount() {
    document.removeEventListener('click', onOutsideClick);
  },
};
