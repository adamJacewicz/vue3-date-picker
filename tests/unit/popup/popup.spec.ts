import { shallowMount, VueWrapper } from '@vue/test-utils';
import Popup                        from '@/components/molecules/popup/popup.vue';
import { clickOutside }             from '@/directives/click-outside';

describe('Popup', () => {
  let wrapper: VueWrapper<any>;
  beforeEach(() => {
   wrapper = shallowMount(Popup, {
      global: {
        directives: {
          clickOutside: clickOutside,
        },
      },
    });
  });

  it('render component with correct class', () => {
    expect(wrapper.classes()).toContain('popup');
  });

  it('show and hide popup on click trigger element', async () => {
    const [ trigger, content ] = wrapper.findAll('.popup__trigger, .popup__content');
    await trigger.trigger('click');
    expect(content.isVisible()).toBe(true);
    await trigger.trigger('click');
    expect(content.isVisible()).toBe(false);
  });

  it('not to hide popup on click popup element', async () => {
    const [ trigger, content ] = wrapper.findAll('.popup__trigger, .popup__content');
    await trigger.trigger('click');
    expect(content.isVisible()).toBe(true);
    await content.trigger('click');
    expect(content.isVisible()).toBe(true);
  });

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
