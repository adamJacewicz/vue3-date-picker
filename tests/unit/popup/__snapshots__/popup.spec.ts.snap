// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup should match snapshot 1`] = `
<div class="popup">
  <div class="popup__trigger"></div>
  <div class="popup__content" style="display: none;">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;
